From 19c26004dcd184003c4f05c12e9907bb32824ceb Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Fri, 26 Mar 2021 18:43:57 -0500
Subject: [PATCH] Fix broken tests: SignSignature is called for tx with invalid
 inputs (#107)

Rebase provenance: this historical change was adapted onto the pinned
Bitcoin ABC base. Source diff is the recorded Lotus V1 replay commit below.

Historical-source-commit: 38367b15c730d884c6091090339feaadd7f80f0c
Recorded-replay-commit: fbdfefb9375c33f6c05591501bf74705beee4910

diff --git a/src/test/orphanage_tests.cpp b/src/test/orphanage_tests.cpp
index e85c51cf60..f40da66c26 100644
--- a/src/test/orphanage_tests.cpp
+++ b/src/test/orphanage_tests.cpp
@@ -159,7 +159,8 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans) {
 
     // This really-big orphan should be ignored:
     for (int i = 0; i < 10; i++) {
-        CTransactionRef txPrev = orphanage.RandomOrphan();
+        CMutableTransaction txPrev =
+            CMutableTransaction(*orphanage.RandomOrphan());
 
         CMutableTransaction tx;
         tx.vout.resize(1);
@@ -167,10 +168,12 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans) {
         tx.vout[0].scriptPubKey =
             GetScriptForDestination(PKHash(key.GetPubKey()));
         tx.vin.resize(2777);
+        txPrev.vout.resize(tx.vin.size());
+        const TxId txid = txPrev.GetId();
         for (size_t j = 0; j < tx.vin.size(); j++) {
-            tx.vin[j].prevout = COutPoint(txPrev->GetId(), j);
+            tx.vin[j].prevout = COutPoint(txid, j);
         }
-        BOOST_CHECK(SignSignature(keystore, *txPrev, tx, 0,
+        BOOST_CHECK(SignSignature(keystore, CTransaction(txPrev), tx, 0,
                                   SigHashType().withForkId()));
         // Re-use same signature for other inputs
         // (they don't have to be valid for this test)
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index 5f4c05e8f9..fe0f004081 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -2413,6 +2413,7 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     pkSingle << ToByteVector(keys[0].GetPubKey()) << OP_CHECKSIG;
     BOOST_CHECK(keystore.AddCScript(pkSingle));
     scriptPubKey = GetScriptForDestination(ScriptHash(pkSingle));
+    txTo.vin[0].prevout = COutPoint(txFrom.GetId(), 0);
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
     scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
@@ -2431,6 +2432,7 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
 
     // Hardest case:  Multisig 2-of-3
     scriptPubKey = GetScriptForMultisig(2, pubkeys);
+    txTo.vin[0].prevout = COutPoint(txFrom.GetId(), 0);
     BOOST_CHECK(keystore.AddCScript(scriptPubKey));
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
